src/Npgsql.GeoJSON/NpgsqlGeoJSONExtensions.cs

```diff
@@ -0,0 +1,61 @@
+﻿#region License
+// The PostgreSQL License
+//
+// Copyright (C) 2018 The Npgsql Development Team
+//
+// Permission to use, copy, modify, and distribute this software and its
+// documentation for any purpose, without fee, and without a written
+// agreement is hereby granted, provided that the above copyright notice
+// and this paragraph and the following two paragraphs appear in all copies.
+//
+// IN NO EVENT SHALL THE NPGSQL DEVELOPMENT TEAM BE LIABLE TO ANY PARTY
+// FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
+// INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS
+// DOCUMENTATION, EVEN IF THE NPGSQL DEVELOPMENT TEAM HAS BEEN ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+// THE NPGSQL DEVELOPMENT TEAM SPECIFICALLY DISCLAIMS ANY WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+// AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE PROVIDED HEREUNDER IS
+// ON AN "AS IS" BASIS, AND THE NPGSQL DEVELOPMENT TEAM HAS NO OBLIGATIONS
+// TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+#endregion
+
+using GeoJSON.Net;
+using GeoJSON.Net.Geometry;
+using Npgsql.GeoJSON;
+using Npgsql.TypeMapping;
+using NpgsqlTypes;
+using System.Data;
+
+namespace Npgsql
+{
+    /// <summary>
+    /// Extension allowing adding the GeoJSON plugin to an Npgsql type mapper.
+    /// </summary>
+    public static class NpgsqlGeoJSONExtensions
+    {
+        /// <summary>
+        /// Sets up GeoJSON mappings for the PostGIS types.
+        /// </summary>
+        /// <param name="mapper">The type mapper to set up (global or connection-specific)</param>
+        /// <param name="option"></param>
+        public static INpgsqlTypeMapper UseGeoJson(this INpgsqlTypeMapper mapper, GeoJSONOption option = GeoJSONOption.None)
+            => mapper
+                .AddMapping(new NpgsqlTypeMappingBuilder
+                {
+                    PgTypeName = "geometry",
+                    NpgsqlDbType = NpgsqlDbType.Geometry,
```

Comment:
Any thoughts about supporting geography as well as geometry?

For the NTS plugin I added this in 8430f7aa97d0c2271a1ebff93f87681f7a841fb5, it's trivial. The wire representation of geometry and geography is the same (WKB), the only difference is what happens when you run PostGIS functions on them; for example, `ST_Distance()` has an overload for geometry and another for geometry - but Npgsql doesn't to care about that.

So supporting geography in Npgsql is basically a matter of allowing users to specify `NpgsqlDbType.Geography` in order to send the geography type OID instead of the geometry one...

Reply:
Maybe we should swap CLR mappings and use `geography` as the default type? Or add an option to choose the default type?

```csharp
public static INpgsqlTypeMapper UseGeoJsonForGeography(this INpgsqlTypeMapper mapper, GeoJSONOption option = GeoJSONOption.None, bool implicitMapping = true)
public static INpgsqlTypeMapper UseGeoJsonForGeometry(this INpgsqlTypeMapper mapper, GeoJSONOption option = GeoJSONOption.None, bool implicitMapping = false);
// or
public static INpgsqlTypeMapper UseGeoJsonForGeometry(this INpgsqlTypeMapper mapper, GeoJSONOption option = GeoJSONOption.None, bool geometryAsDefault = false);
```

I think that the first one is better because it registers only one factory which will fetch authorities when constructing a handler.

Reply:
I think that `geography` definitely shouldn't be the default type... `geography` is much more limited in terms of operations you can execute (because it's much more complicated), all operations cost much more in terms of CPU (again, much more complicated :)), and it's only needed when working on a large or global scale (see [these notes in the PostGIS docs](https://postgis.net/docs/manual-2.3/using_postgis_dbmanagement.html#PostGIS_Geography)). Npgsql has actually never supported `geography` with the legacy built-in types and very few people complained, I think it's pretty rare.

I guess a bool flag to `UseGeoJSON()` for using `geography` by default could be a good idea... By the way, if somebody wants to map everything to `geography` instead of `geometry`, it's pretty trivial to write some code in the models' `OnModelConfiguring()` which finds all spatial properties and sets their column type to geography. But if we can help out with a flag, why not.

If you add this flag, can you submit a PR doing the same for the NTS plugin?

Reply:
Okay, but what is about fetching authorities? Each factory (for `geometry` and `geography`) will request records from `spatial_ref_sys` when `GeoJSONOptions.ShortCRS` or `GeoJSONOptions.LongCRS` is specified. I can fix it, but it would be tricky (linked factories).

I think that type handling could be optimized by caching handlers, because now Ngpsql queries types every time connection gets open and allocates memory for their handlers. If someone wants to reload handlers and their data, then he will use `ReloadTypes`.

Reply:
I don't much about the authority records... Wouldn't the flag only affect the (default) mapping for the CLR types? In other words, in the current NTS plugin, CLR `Point` is (by default) mapped to to the geometry mapping, not geography (i.e. an instance of `NetTopologySuiteGeometryTypeMapping`). If you make it mapped to the `NetTopologySuiteGeographyTypeMapping` instance instead (when the flag is specified), isn't that enough?

Reply:
The flag isn't related to authorities.